eip-0008.md

```diff
@@ -0,0 +1,479 @@
+with (i.e. sending $10 and pay $100 for the gas doesn't have economic sense).
+
+### Ergo Transaction Validation With JIT Costing
+
+Here we describe an extension to the existing Ergo protocol, which allows to prevent execution
+of scripts beyond allotted limits. The extension doesn't require changes in the
+serialization format. Moreover, simple _transfer transactions_ are processed the same way so
+that 1) most existing applications (like exchanges) will not require any changes and 2)
+the new protocol doesn't incure additional overhead for simple transfer transactions.
+The cost calculation of ErgoTrees is implemented as it is defined in [the next
+section](#jit-costing-of-ergotree).
+
+On a high level, the new protocol for transaction creation, signing and validation is
+described as a series of steps. The anatomy of the transactions which are created is shown
+in the figure below.
+
+ 1) A transaction sender performs cost estimation as part of transaction signing.
+Transaction signing involves: 1) execution of the scripts for all the input boxes and 2)
+execution of sigma protocols to generate proofs. This happens using the current context of
+the blockchain at the time of transaction signing, in particular current blockchain height
+is used to execute scripts and may influence the outcome of the scripts. [JIT
+costing](#jit-costing-of-ergotree) is performed during 1) and 2) steps resulting in the
+value `estCost` to be produced. Note, if scripts execution doesn't depend on the current
+blockchain parameters (like `HEIGHT`, `headers` etc.) then the `estCost` accurately
+predicts execution costs in any context (i.e. `estCost == actCost`, see below). This is
+because script execution depends only on `tx` content (which is immutable) and on nothing
+else. For majority of scripts this is the case.
+
+ 2) The transaction sender must decide on the `costPrice` and `costLimit` values. The
+ `costLimit` parameter is used by the miners to reject over-expensive transactions and
+ should be greater or equal to `estCost` estimated during `tx` signing. The sender must
+ put at least `feeLimit = costLimit * costPrice` NERGs (NanoERGs) in the miner's fee
+ output box (see `feeOut` in the figure).
+The sender in addition puts the following values in the registers of `feeOut` box ([see
+footnote 2](#footnotes)):
+    - R4: a `Long` value of `costPrice` which is equal to the number of NanoERG to be
+    paid per unit of cost.
+    - R5: a `senderPk` public key belonging to the sender. The ownership is proved
+    during `tx` signing by spending at least one box protected by the `p2pk(senderPk)`
+    contract.
+    - R6: an interval of blockchain heights in which the transaction is valid, a pair
+    of two Int values.
+ 
+3) The transaction sender needs to convince miners that he has enough coins to pay for
+transaction fees up to `feeLimit`. He can do it by showing enough boxes protected by
```

Comment:
no need to convince if there's fee output already

Reply:
the key point here is 
"_He can do it by showing enough boxes _protected by
`p2pk(senderPk)` script_ corresponding to the public key stored in `feeOut.R5` register._"

Reply:
why the fee can't be paid without all these complexities with inputs?

Reply:
Well, to avoid the attack you mentioned above. This is exactly the mechanism to avoid it.